src/librustc_typeck/check/mod.rs

```diff
@@ -1064,6 +1072,15 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
         fcx.check_pat_walk(&arg.pat, arg_ty,
             ty::BindingMode::BindByValue(hir::Mutability::MutImmutable), true);
 
+        // If any of a function's parameters have a type that is uninhabited, then it
+        // may never be called (because its arguments cannot be constructed). Therefore,
+        // it must always diverge.
+        if fcx.tcx.features().better_divergence_checking {
+            if arg_ty.conservative_is_uninhabited() {
+                fcx.diverges.set(fcx.diverges.get() | Diverges::Always);
+            }
+        }
```

Comment:
it's definitely an interesting "judgement call" whether we want to issue a warning in this case

Reply:
(probably?)